runtime/signals.c

```diff
@@ -350,56 +349,56 @@ value caml_do_pending_actions_exn(void)
   Caml_state->action_pending = 0;
 
   /* Call signal handlers first */
-  value exn = caml_process_pending_signals_exn();
-  if (Is_exception_result(exn)) goto exception;
+  caml_result result = caml_process_pending_signals_result();
+  if (result.is_exception) goto exception;
 
 #if 0
   /* Call memprof callbacks */
-  exn = caml_memprof_handle_postponed_exn();
-  if (Is_exception_result(exn)) goto exception;
+  result = caml_memprof_handle_postponed_result();
+  if (result.is_exception) goto exception;
 #endif
 
   /* Call finalisers */
-  exn = caml_final_do_calls_exn();
-  if (Is_exception_result(exn)) goto exception;
+  result = caml_final_do_calls_result();
+  if (result.is_exception) goto exception;
 
   /* Process external interrupts (e.g. preemptive systhread switching).
      By doing this last, we do not need to set the action pending flag
      in case a context switch happens: all actions have been processed
      at this point. */
   caml_process_external_interrupt();
 
-  return Val_unit;
+  return result;
 
 exception:
   /* If an exception is raised during an asynchronous callback, then
      it might be the case that we did not run all the callbacks we
      needed. Therefore, we set [Caml_state->action_pending] again in
      order to force reexamination of callbacks. */
   caml_set_action_pending(Caml_state);
-  return exn;
+  return result;
 }
 
-value caml_process_pending_actions_with_root_exn(value root)
+caml_result caml_process_pending_actions_with_root_result(value root)
 {
   if (caml_check_pending_actions()) {
     CAMLparam1(root);
-    value exn = caml_do_pending_actions_exn();
-    if (Is_exception_result(exn)) CAMLreturn(exn);
+    caml_result result = caml_do_pending_actions_result();
+    if (result.is_exception) CAMLreturnT(caml_result, result);
     CAMLdrop;
   }
-  return root;
+  return Result_value(root);
 }
 
 value caml_process_pending_actions_with_root(value root)
 {
-  return caml_raise_if_exception(
-    caml_process_pending_actions_with_root_exn(root));
+  return caml_run_result(
+    caml_process_pending_actions_with_root_result(root));
 }
 
-CAMLexport value caml_process_pending_actions_exn(void)
+CAMLexport caml_result caml_process_pending_actions_result(void)
 {
-  return caml_process_pending_actions_with_root_exn(Val_unit);
+  return caml_process_pending_actions_with_root_result(Val_unit);
 }
```

Comment:
Too much wrapping going on here... I think this function is the same as `caml_do_pending_actions_result`.

Reply:
This is not just `caml_do_pending_actions_result`, as there is a check on `caml_check_pending_actions` first. I inlined/streamlined the code -- and took the opportunity to write a small bit of documentation on the `with_root` variants. This new commit will need a review because it is less of a mechanical refactoring than the rest of the patch.